src/python/grpcio/grpc/_cython/_cygrpc/grpc_eventlet.pyx.pxi

```diff
@@ -0,0 +1,691 @@
+        socket_wrapper.users.remove(current)
+    except (IOError, OSError) as exc:
+        socket_wrapper.users.remove(current)
+        socket_wrapper.write_cb(socket_wrapper.c_socket,
+                                socket_error('send', str(exc)))
+        socket_wrapper.call_close_cb()
+    eventlet_kick_poll()
+
+
+cdef void eventlet_socket_write(grpc_custom_socket* socket,
+                                grpc_slice_buffer* buffer,
+                                grpc_custom_write_callback cb) with gil:
+    cdef char* start
+    sw = <ESocketWrapper>socket.impl
+    sw.write_cb = cb
+    data = bytearray()
+    for i in range(buffer.count):
+        start = grpc_slice_buffer_start(buffer, i)
+        length = grpc_slice_buffer_length(buffer, i)
+        data.extend(<bytes>start[:length])
+    eventlet_spawn(eventlet_socket_write_async, sw, data)
+
+
+cdef eventlet_socket_read_async(ESocketWrapper socket_wrapper):
+    cdef char* buff_char_arr
+    try:
+        current = e_greenlet.getcurrent()
+        socket_wrapper.users.append(current)
+        buff_str = socket_wrapper.socket.recv(socket_wrapper.len)
+        buff_char_arr = buff_str
+        string.memcpy(<void*>socket_wrapper.c_buffer, buff_char_arr,
+                      len(buff_str))
+        socket_wrapper.read_cb(socket_wrapper.c_socket, len(buff_str),
+                               <grpc_error*>0)
+        socket_wrapper.users.remove(current)
+    except (IOError, OSError) as exc:
+        socket_wrapper.users.remove(current)
+        socket_wrapper.read_cb(<grpc_custom_socket*>socket_wrapper.c_socket,
+                               -1, socket_error('recv', str(exc)))
+        socket_wrapper.call_close_cb()
+    eventlet_kick_poll()
+
+
+cdef void eventlet_socket_read(grpc_custom_socket* socket, char* buffer,
+                               size_t length,
+                               grpc_custom_read_callback cb) with gil:
+    sw = <ESocketWrapper>socket.impl
+    sw.read_cb = cb
+    sw.c_buffer = buffer
+    sw.len = length
+    eventlet_spawn(eventlet_socket_read_async, sw)
+
+
+cdef grpc_error* eventlet_socket_getpeername(grpc_custom_socket* socket,
+                                             const grpc_sockaddr* addr,
+                                             int* length) with gil:
+    cdef grpc_resolved_address c_addr
+
+    peer = (<ESocketWrapper>socket.impl).socket.getpeername()
+    hostname = str_to_bytes(peer[0])
+    grpc_string_to_sockaddr(&c_addr, hostname, peer[1])
+    string.memcpy(<void*>addr, <void*>c_addr.addr, c_addr.len)
+    length[0] = c_addr.len
+    return <grpc_error*>0
+
+
+cdef grpc_error* eventlet_socket_getsockname(grpc_custom_socket* socket,
+                                             const grpc_sockaddr* addr,
+                                             int* length) with gil:
+    cdef grpc_resolved_address c_addr
+    if (<ESocketWrapper>socket.impl).socket is None:
+        peer = ('0.0.0.0', 0)
```

Comment:
This would happen if we call socket init and then call get peer name.
I don't know if the gRPC core code can actually end up making these calls like that to a custom IO manager, but since Python doesn't support `AF_UNSPEC` sockets and we can have a `None` value in the socket I think it's better to take it into account as a possibility.